### ISSUE TITLE
Add Bollinger Bands Indicator

### ISSUE THREAD

**Issue by Bartoshko** [greyblake/ta-rs#6]
I think that Bollinger Bands Indicator would be a nice to have feature.
I see implementation on this indicator as:
 - struct that will extend moving average struct by using it as part of self data struct,
 - will share common project trait,
 - next function will return tuple of (moving average: i32, upper_bound: i32, lower_bound: i32 ) or corresponding data struct,
 - new function will take (n: i32, k: i32) where n is buffer size and k is standard deviation multiplier that multiply itself by standard deviation to indicate upper and lower bound according to Bollinger Band formula,

**Comment by greyblake** [comment]
Hi! Thanks for your proposal. Now I see what was the reason for standard deviation :) 

> struct that will extend moving average struct by using it as part of self data struct,
It's probably obvious, but could you name the indicator `BollingerBands` ?
Would you like to use Simple Moving Average or Exponential Moving Average? (it looks like there are different invariant of BollingerBands). I usually tend to use EMA but don't have strong arguments and preferences.

> next function will return tuple of (moving average: i32, upper_bound: i32, lower_bound: i32 ) or corresponding data struct,

Returning data struct is a good idea, because it makes things more clear. What would be a good name for such data struct? `BollingerBandValue` (or ...Values) ? (it is important to pick a good name here, I'd like to extend your idea, and apply the same naming pattern to MACD, because returning tuple is confusing).
What would be the field names? 

> new function will take (n: i32, k: i32) where n is buffer size and k is standard deviation multiplier that multiply itself by standard deviation to indicate upper and lower bound according to Bollinger Band formula,

You mean that `BollingerBands::new` would take `(n: i32, k: i32)` ? If so, then it looks good to me.

Regarding standard deviation: I would suggest implementing as a separate function that can be tested in isolation and then using this function within `next` function. Or alternatively finding a small crate that could provide this function (if you decide to go this way, please let me validate the crate first).

Again thanks for your initiative. Please don't take every my comment as a final decision. If you see that something here does not make sense or you see a better approach to do something (e.g. better names, etc.) I would be happy to discuss such things! =)

Thank you.

**Comment by greyblake** [comment]
Ah, important comment. Usually I verify my implementation and test data against indicators at `investing.com`.
Could you please do the same? 
https://www.investing.com/equities/amazon-com-inc-chart

Here are screenshot instructions:

![image](https://user-images.githubusercontent.com/113512/54089315-189b4700-4368-11e9-965b-8540ab7de784.png)
![image](https://user-images.githubusercontent.com/113512/54089323-251f9f80-4368-11e9-8a56-c395f09a1864.png)
![image](https://user-images.githubusercontent.com/113512/54089332-336dbb80-4368-11e9-94f0-f0e415226a76.png)
![image](https://user-images.githubusercontent.com/113512/54089336-3cf72380-4368-11e9-94f1-9f75af9d1e81.png)

**Comment by Bartoshko** [comment]
Hi, Thanks for quick replay and all suggestions. I don't have any objections and I think this is very nice design pattern to follow. I will implement returning type for `next` as `struct` of suggested name. I am not good at naming things, and sometimes I fall in to a trap of enigmatic names that works only for me, so fell free to point it out, and by the way; thanks for descriptions. First I will bring some code as a brief prototype to discuss, but have only one request, which is that I do not always have systematic time for contribution, so if something would go few days without response please 'forgive' :). But most of the time I stay very connected to the world. To be consistent in this project, please if you have any rules for example branch naming, please let me know. Thanks for all. In touch. I hope weather in Berlin is nice, in Gdansk We are hosting Scandinavian wind.

**Comment by Bartoshko** [comment]
Regarding to field names, I will be back with proposal asap.

**Comment by Bartoshko** [comment]
What do you think to use `enum` for example: 
```
enum AverageType {
    SMA,
    EMA,
}
```
as flexible setup of which calculation of average library user would be happy to use?

**Comment by greyblake** [comment]
I would rather use a generic for this purpose... But let's leave it for the future.
Currently we can go with one simple option.


P.S. 
In Berlin we have rain...
I was few times in Gdansk, really beautiful city! Sometimes I have to travel through it (I have relatives in Kaliningrad :) )

P.S.2 
I may not respond for few days.. But usually I try to keep things on track ;)

**Comment by Bartoshko** [comment]
Lets go with EMA than. 
P.S. I have heard that Kaliningrad is a beautiful city, I have got a friend there, but now lives in Murmansk.

**Comment by Bartoshko** [comment]
Proposal:
```rust
use std::fmt;

use crate::errors::*;
use crate::{Close, Next, Reset};
use crate::indicators::ExponentialMovingAverage;

/// A Bollinger Bands (BB).
/// (BB).
/// It is a type of infinite impulse response filter that calculates Bollinger Bands using Simple Moving Average or Exponential Moving Average depends on settings.
/// The Bollinger Badns are represented by Average (EMA or EWMA) and standard deviaton that is moved 'k' times away in both directions of calculated average value.
/// 
/// # Formula
///
/// Bollinger Bands are calculated based on EWMA or SMA combined with Standard Deviaiation(SD).
/// See EWMA and SMA doumentation.
///
/// ![SD formula](https://wikimedia.org/api/rest_v1/media/math/render/svg/8715d659116bec91d48762b9e1f3d9aed36fc028)
///
/// Where:
///
/// * _SD<sub>s</sub>_ - is value of standard deviation for N given probes.
/// * _SD<sub>ٍ-</sub>_  - is the mean value of observation.
/// * _SD<sub>N</sub>_ - is number of probes in observation.
/// * _SD<sud>xi</sub>_ - is i-th observed value from N elements observation.
///
/// and then BB is composed as:
///
///    Middle Band = N simple moving average (SMA) or Exponential Moving Average (EWMA).
///    Upper Band = N * ((SMA or EWMA) + SD of observation * multipler (usually 2.0))
///    Lower Band = N * ((SMA or EWMA) - SD of observation * multipler (usually 2.0))
///
/// # Example
///
///```
/// use ta::indicators::{BollingerBands, BandsType, BandsPayload};
/// use ta::Next;
///
/// let mut bb = BollingerBands::new(BandsType::SMA, 20, 2.0_f64).unwrap();
///
///
/// assert_eq!(bb.next(2.0), BandsPayload {
///     average: 2.0_f64,
///     upper: 2.0_f64,
///     lower: 2.0_f64,
/// });
///
/// assert_eq!(bb.next(4.0), BandsPayload {
///     average: 3.0_f64,
///     upper: 0.0_f64,
///     lower: 4.0_f64,
/// });
/// ```
///
/// # Links
///
/// * [Bollinger Bands, Wikipedia](https://en.wikipedia.org/wiki/Bollinger_Bands)
///


```

**Comment by greyblake** [comment]
@Bartoshko 
It looks like the proposal does not count our latest discussion:
* `BandsType` - we agreed to use EMA instead (for now at least)
* `BandsPayload` - I am not comfortable with this name. In my mind it resonate a lot with HTTP communication. `BollingerBandsValue` may sound also strange.. But I'd like to develop some common convention for such structs. I think something like `<indicator name><some word>`. I see `Value` as `<some word>` at the moment. But could be something else: `Values`, `Data`, `Output`, etc ? I am open for some flexibility but I'd like to ensure that we pick some option that would be suitable for other indicators (like MACD), so we can be consistent. 
As I write it.. I realize that I like `Output` more than `Value`, because it would kind of reference `Output` associated type of the `Next<T>` trait.
So what do you think about `BollingerBandsOutput` ?:)

There are also some typos in the doc. But general idea looks good :+1:

**Comment by Bartoshko** [comment]
Ok, than. All set: 
 - EMA for average, 
 - BollingerBandsValue is returned type from BoolingerBand indicator struct.

Looks like a lot talk but as has been said better to resolve all unknowns from beginning.

Thanks for all tips. :+1:

**Comment by greyblake** [comment]
> BollingerBandsValue is returned type from BoolingerBand indicator struct.

Sorry, I probably should not write a novel in my comments. 
I would prefer `BollingerBandsOutput` instead of `BollingerBandsValue`.

**Comment by Bartoshko** [comment]
Sorry, I wasn't referring to Your comment, more to my questions and preparation phase proposal. Respect for all your suggestion, and thx.

**Comment by greyblake** [comment]
Bollinger Bands are released in version 0.1.2 https://docs.rs/ta/0.1.2/ta/indicators/struct.BollingerBands.html
